### ISSUE TITLE
Can't be tested with Jest

### ISSUE THREAD

**Issue by nirsky** [elastic/require-in-the-middle#50]
Seems like Jest magic on the require function is clashing with require-in-the-middle.
Not sure if there's a way to overcome this.
Decided to open this issue to discuss this, maybe someone can share a solution.

Related:
* https://overengineered.thoughts.lt/testing-imports/
* https://github.com/facebook/jest/issues/7722

**Comment by astorm** [comment]
Hello @nirsky, thanks for bringing this up.  To start the conversation, two questions

1. Do you have an example of some jest code where require-in-the-middle doesn't work like you expect?
2. Do you happen to know what, specifically, "jest magic on the require function" is and/or where in jest they're applying this magic?

**Comment by nirsky** [comment]
Hey, sorry for the late reply.
I encountered this issue while writing tests for [plugins](https://github.com/aspecto-io/opentelemetry-ext-js) I wrote for [opentelemetry](https://github.com/open-telemetry/opentelemetry-js).
opentelemetry uses require-in-the-middle to know when a library was required, and patch it upon requiring it.

Jest is overriding the `require` to create an isolated environment for each test suite, mocking and JIT (https://github.com/facebook/jest/issues/7722).  
I suspect those 2 collide.  
In my case I had to migrate all the tests to use mocha.

**Comment by watson** [comment]
You're correct regarding jest vs require-in-the-middle. Jest uses the `vm` core module to run tests. I've spent a lot of time to see if we could somehow be notified of calls to `require` from within those vm's, but so far no luck. It should be technically possible, but jest is very complicated, so I've not been able to figure out what change I need to make to jest in order for it to work. So for now I've tabled the efforts as it was taking too much time.

**Comment by shulamit74** [comment]
I have a similar issue - We have a `graphql` app that uses `@graphile-contrib/pg-many-to-many`.
We test it with supertest and jest.
I was able to run the tests, but got a "Jest did not exit one second after the test run has completed" warning.
After graying out the tests, I saw this error: `TypeError: require(...) is not a function` [In this line](https://github.com/graphile-contrib/pg-many-to-many/blob/715e84fb3b021e00bbfc20031bd8a3c5e51c4050/index.js#L30) in an inline require.
The require at that point is `[Function: bound requireModuleOrMock]`. Is the inline require an issue ? Any way to overcome this?
